synfig-studio/src/gui/dialogs/canvasresize.cpp

```diff
@@ -0,0 +1,285 @@
+	height->signal_changed().connect(sigc::bind<Gtk::SpinButton *>(
+			sigc::mem_fun(*this, &CanvasResize::on_spinbutton_updated), height));
+	height->signal_value_changed().connect(
+			sigc::mem_fun(*this, &CanvasResize::on_height_changed));
+	canvas_only->signal_toggled().connect(
+			sigc::mem_fun(*this, &CanvasResize::on_canvas_only_toggled));
+	canvas_label->add_events(Gdk::POINTER_MOTION_MASK);
+	canvas_label->signal_motion_notify_event().connect(
+			sigc::mem_fun(*this, &CanvasResize::on_canvas_label_motion_event));
+	canvas_label->signal_button_press_event().connect(
+			sigc::mem_fun(*this, &CanvasResize::on_canvas_label_button_event));
+	toggle_ratio_wh->signal_toggled().connect(
+			sigc::mem_fun(*this, &CanvasResize::on_wh_ratio_toggled));
+
+	// Action buttons
+	signal_response().connect(
+			sigc::mem_fun(*this, &CanvasResize::on_action_signal_response));
+}
+
+void CanvasResize::on_rend_desc_changed()
+{
+	rend_desc = canvas_interface->get_canvas()->rend_desc();
+
+	refresh_content();
+}
+
+void CanvasResize::on_action_signal_response(int response_id)
+{
+	auto rend_desc_old = canvas_interface->get_canvas()->rend_desc();
+	auto width_old     = rend_desc_old.get_w();
+	auto height_old    = rend_desc_old.get_h();
+	auto was_toggled   = rend_desc_old.get_flags() & RendDesc::LINK_IM_ASPECT;
+	auto width_new     = rend_desc.get_w();
+	auto height_new    = rend_desc.get_h();
+	auto is_toggled    = rend_desc.get_flags() & RendDesc::LINK_IM_ASPECT;
+
+	switch (response_id) {
+	case 1:
+		// Advanced was pressed
+		canvas_properties.present();
+	case 2:
+		// Cancel was pressed
+	case Gtk::RESPONSE_DELETE_EVENT:
+		// Close was pressed - title bar
+		rend_desc = canvas_interface->get_canvas()->rend_desc();
+		refresh_content();
+		break;
+	case 3:
+		// OK was pressed
+		if (
+				width_new  != width_old  ||
+				height_new != height_old ||
+				is_toggled != was_toggled
+		) {
+			// Set new pixel ratio before setting new rend_desc
+			rend_desc.set_pixel_ratio(width_new, height_new);
+			canvas_interface->set_rend_desc(rend_desc);
+		}
+	}
+
+	hide();
+}
+
+void CanvasResize::on_width_changed()
+{
+	if (update_lock) return;
+	UpdateLock lock(update_lock);
+
+	rend_desc.set_w(width->get_value_as_int());
+
+	refresh_content();
+}
+
+void CanvasResize::on_height_changed()
+{
+	if (update_lock) return;
+	UpdateLock lock(update_lock);
+
+	rend_desc.set_h(height->get_value_as_int());
+
+	refresh_content();
+}
+
+bool CanvasResize::on_canvas_label_motion_event(GdkEventMotion *event)
+{
+	// Handle focus-hover-like event for canvas_only widget here
+
+	return false;
+}
+
+bool CanvasResize::on_canvas_label_button_event(GdkEventButton *event)
+{
+	if (event->button == 1 && event->type == GDK_BUTTON_PRESS) {
+		canvas_only->set_active(!canvas_only->get_active());
+		return true;
+	}
+
+	return false;
+}
+
+void CanvasResize::on_canvas_only_toggled()
+{
+	auto is_active  = canvas_only->get_active();
```

Comment:
why not just bool? :)

Reply:
Habit LOL. I just love C++ auto type deduction hehe